Add 5 new options: -f -ff -oo -r -e and update README.md

junyuan-chen opened the pull request:
--fragment, -f: output fragment of LaTeX code used inside a tabular environment
--fragment-skip-header, -ff: shortcut of -k 1 -n -f
--separate-outfiles, -oo: save each table into a separate tex file
--replace, -r: replace any existing output file specified in -o with the new output
--no-escape, -e: do not escape special LaTeX characters

Two new examples are added into README.md to illustrate the usage of the new options.
The fragment option is useful when the column titles are complicated.
In that scenario, it can be more convenient to input the table fragment into a main tex file and edit the column titles from the main file.

junyuan-chen committed:
Add 5 new options: -f -ff -oo -r -e and update README.md

--fragment, -f: output fragment of LaTeX code used inside a tabular environment
--fragment-skip-header, -ff: shortcut of -k 1 -n -f
--separate-outfiles, -oo: save each table into a separate tex file
--replace, -r: replace any existing output file specified in -o with the new output
--no-escape, -e: do not escape special LaTeX characters

narimiran commented:
Thank you for your contribution @junyuan-chen! I was on holidays so I couldn't respond earlier.

I'll take a look at your changes/additions shortly.

narimiran commented on tably.py:
```diff
@@ -19,6 +22,7 @@
 
 LABEL = '\n{indent}\\label{{{label}}}'
 CAPTION = '\n{indent}\\caption{{{caption}}}'
+ESCAPE = True
```
These things defined here should be constants, but later on you modify `ESCAPE`.

narimiran commented on tably.py:
```diff
@@ -66,8 +80,24 @@ def run(self):
         all_tables = []
+
+        if self.no_escape:
+            global ESCAPE
```
I would like to avoid `global` as it is usually a sign of code smell.

Can this be done some other way? E.g. `-e` option which is `True` by default, and then use `self.no_escape` later on when needed.

narimiran commented on tably.py:
```diff
@@ -78,16 +108,21 @@ def run(self):
+        elif self.separate_outfiles is not None: # if -oo is passed
+            outs = self.separate_outfiles
+            if len(outs) == 0:
+                outs = [ file[:-4]+'.tex' if file[-4:]=='.csv' else file+'.tex' for file in self.files]
```
What if there is a file with an extension which is not `.csv`? This would make it `filename.ext.tex`.

narimiran commented on tably.py:
```diff
@@ -78,16 +108,21 @@ def run(self):
+            if len(outs) == 0:
+                outs = [ file[:-4]+'.tex' if file[-4:]=='.csv' else file+'.tex' for file in self.files]
+            elif len(outs) != len(self.files):
+                print('WARNING: Number of .cvs files and number of output files do not match')
```
typo: `.cvs` -> `.csv`

narimiran commented on tably.py:
```diff
@@ -120,15 +155,29 @@ def create_table(self, file):
+            return '\n'.join((header, content, footer))
+        else:
+            return content
+    def save_each_table(self, file, out):
```
Nitpicking, but since a single table is passed, a better function name reflecting that would be `save_single_table` or something like that.

narimiran commented on tably.py:
```diff
@@ -281,6 +346,28 @@ def arg_parser():
+        '-ff', '--fragment-skip-header',
+        action='store_true',
+        help='Equivalent to passing -k 1 -n -f '
+             '(suppress header when they are on the first row of .csv and pass -f) '
```
Nitpicking:

Whitespace at the end of a line is needed when there is a line/text following it. When it is the final line, it should end with a period and no whitespace after it.

narimiran reviewed (changes requested):
I've taken a quick look at your changes and I like the most of it!

But there are some small details I would like to see fixed/changed before merging.

junyuan-chen committed:
Avoid modifying global variable and make other improvements

- Variable ESCAPE is avoided by passing an additional argument to the relevant functions
- Reorganize Tably.run() to make the code more readable
- Use os module to handle file paths
- Allow a directory to be passed after -oo option
- Update README.md

junyuan-chen commented:
Thank you for your comment! @narimiran 

I have modified the code based on your suggestions.

narimiran commented on tably.py:
```diff
@@ -163,9 +218,11 @@ def format_alignment(align, length):
 
 
-def get_units(units):
+def get_units(units, no_escape):
```
Since we're passing `self.units` and `self.no_escape` (see line 144), and it is not used/called outside the Tably class, maybe we should convert this function to a class method.

What do you think?

junyuan-chen replied:
Well, it's true that `get_units` is not called outside the class. But, instead of directly making it a class method, I am thinking about doing something different in the `__init__` method of `Tably` class to avoid passing `no_escape` and doing if-else branching inside the methods every time they are called.

junyuan-chen replied:
I have just pushed the new way to handle escaping/no-escaping. I am not sure whether it is a good practice to set a class attribute as a function in Python. But, it doesn't seem to bring harm and does make the code looks better.

junyuan-chen committed:
Improve handling of no_escape

- Define tex_str method in __init__ of Tably class based on --no-escape option
- Make get_units() and create_row() methods of Tably class

narimiran commented on tably.py:
```diff
@@ -67,10 +67,13 @@ def __init__(self, args):
         self.fragment = args.fragment
         self.fragment_skip_header = args.fragment_skip_header
         self.replace = args.replace
+        if args.no_escape:
```
Nice idea!

I would write this as a ternary operator with no-op lambda, so there's no need to define `do_not_escape` function which does nothing:

```python
self.tex_str = escape if not args.no_escape else lambda x: x
```

junyuan-chen replied:
OK, code updated.

narimiran commented on tably.py:
```diff
@@ -156,6 +159,12 @@ def create_table(self, file):
+        """Creates a row based on `line` content"""
+        return r'{indent}{indent}{content} \\'.format(
+             indent=indent,
+             content=' & '.join(self.tex_str(line)))
```
I agree with you — much better/cleaner than the previous version of `create_row`!

junyuan-chen committed:
Simplify expression and update docstrings

- One-line assignment of self.tex_str in Tably class
- Update docstrings for Tably class
